offchain/danaswap-api/src/DanaSwap/Api.purs

```diff
@@ -9,30 +17,119 @@ import Contract.Prelude
 import Contract.Utxos (getUtxo)
-import Contract.Value (CurrencySymbol, adaToken, mpsSymbol, scriptCurrencySymbol)
+import Contract.Value (CurrencySymbol, TokenName, adaToken, mkTokenName, mpsSymbol, scriptCurrencySymbol, symbols, valueOf)
 import Contract.Value as Value
 import Ctl.Util (buildBalanceSignAndSubmitTx, getUtxos, waitForTx)
 import DanaSwap.CborTyped (configAddressValidator, liqudityTokenMintingPolicy, poolAddressValidator, poolIdTokenMintingPolicy, simpleNft)
 import Data.BigInt as BigInt
 import Data.List (head)
-import Data.Map (keys)
+import Data.Map (Map, keys)
 import Data.Map as Map
 import Data.Set (toUnfoldable)
+import Effect.Exception (throw)
 
 type Protocol =
   { configUtxo :: TransactionInput
-  , poolVal :: Validator
+  , poolAdrVal :: Validator
   , liquidityMP :: MintingPolicy
   , poolIdMP :: MintingPolicy
   }
 
+type PoolId = TokenName
+
+getAllPools :: Protocol -> Contract () (Map TransactionInput TransactionOutputWithRefScript)
+getAllPools protocol@{ poolAdrVal } =
+  getUtxos (scriptHashAddress $ validatorHash poolAdrVal)
+    <#> Map.filter (hasNft protocol)
+
+getPoolById :: Protocol -> PoolId -> Contract () (TransactionInput /\ TransactionOutputWithRefScript)
+getPoolById protocol@{ poolIdMP } token = do
+  pools <- getAllPools protocol
+  cs <- liftContractM "Failed to get the currency symbol for the protocols mintingPolicy" $ mpsSymbol $ mintingPolicyHash poolIdMP
+  let valid = Map.filter (\vault -> valueOf (unwrap (unwrap vault).output).amount cs token > BigInt.fromInt 0) pools
+  case Map.toUnfoldableUnordered valid of
+    [] -> liftEffect $ throw "no pools with that ID"
+    [ vault ] -> pure vault
+    _ -> liftEffect $ throw "more than one pool with the same ID, this is really bad"
+
+hasNft :: Protocol -> TransactionOutputWithRefScript -> Boolean
+hasNft { poolIdMP } out = case (mpsSymbol $ mintingPolicyHash poolIdMP) of
+  Nothing -> false -- protocol was invalid
+  Just cs -> cs `elem` (symbols $ (unwrap (unwrap out).output).amount)
+
+-- TODO this is a placeholder implementation
+depositLiquidity :: Protocol -> PoolId -> Contract () Unit
+depositLiquidity protocol@{ poolAdrVal, liquidityMP, poolIdMP } poolID = do
+  (poolIn /\ poolOut) <- getPoolById protocol poolID
+  poolIdCs <- liftContractM "hash was bad hex string" $ mpsSymbol $ mintingPolicyHash poolIdMP
+  let idNft = Value.singleton poolIdCs poolID one
+  void $ waitForTx (scriptHashAddress $ validatorHash poolAdrVal) =<<
+    buildBalanceSignAndSubmitTx
+      ( Lookups.unspentOutputs (Map.singleton poolIn poolOut)
+          <> Lookups.mintingPolicy liquidityMP
+          <> Lookups.validator poolAdrVal
+      )
+      ( Constraints.mustSpendScriptOutput
+          poolIn
+          (Redeemer $ toData unit)
+          <> Constraints.mustMintCurrencyWithRedeemer
+            (mintingPolicyHash liquidityMP)
+            (Redeemer $ List [ toData poolID, Constr one [] ])
+            poolID
+            (BigInt.fromInt 10)
```

Comment:
Could you move this 10 to a named variable? It's not immediately clear to me what it means.

Reply:
It's just an arbitrary placeholder amount until there's a real pool validator for deposit Liquidity to interact with.